[ExecuTorch] Add broadcast support for optimized add op

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #8257
* #8256
* #8255
* __->__ #8205

Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc @larryliu0820 @manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

kimishpatel committed:
[Executorch] Refactor op_mul's broadcasting utils

Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
[ExecuTorch] Add broadcast support for optimized add op

Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

swolchok commented on kernels/optimized/cpu/op_add.cpp:
```diff
@@ -140,41 +140,31 @@ Tensor& opt_add_out(
+            ElementwiseOptimizedPath::kBroadcastLastDimReverseArguments ||
+        selected_optimized_path ==
+            ElementwiseOptimizedPath::kBroadcastNdByNdReverseArguments) {
+      // This behavior is a bit confusing.
```
I don't understand what's confusing here; there is an argument that should be scaled by alpha_val, we have to scale the right one. definitely don't think handle_broadcast_elementwise should be coupled to the specific op.

kimishpatel replied:
problem is this. All the reverse arg stuff has specifically different handlking inside `handle_broadcast_elementwise`. But that handling, inside `handle_broadcast_elementwise`, to work this change is necessary which makes them coupled and fragile

kimishpatel replied:
I guess confusing is not the right word here though.

swolchok commented on kernels/optimized/cpu/op_mul.cpp:
```diff
@@ -130,8 +130,12 @@ Tensor& opt_mul_out(
   } else if (selected_optimized_path != ElementwiseOptimizedPath::kNone) {
-    auto mul_lambda = [](auto x, auto y) { return x * y; };
-    return torch::executor::handle_broadcast_elementwise(
+    // Reason for using alpha:
```
missing rest of comment after the colon

swolchok commented on kernels/optimized/cpu/op_mul.cpp:
```diff
@@ -130,8 +130,12 @@ Tensor& opt_mul_out(
-    return torch::executor::handle_broadcast_elementwise(
+    // Reason for using alpha:
+    auto mul_lambda = [](auto x, auto y, auto alpha) {
+      (void)alpha;
```
https://en.cppreference.com/w/cpp/language/attributes/maybe_unused

kimishpatel replied:
thank you :)

swolchok commented on kernels/optimized/cpu/binary_ops.h:
```diff
@@ -48,8 +49,38 @@ enum class ElementwiseOptimizedPath {
+
 namespace internal {
 
+template <BinaryOpType op_type>
+struct BinaryOpTypeName;
+
+template <>
+struct BinaryOpTypeName<BinaryOpType::kAdd> {
+  static constexpr char kName[] = "add.out";
+};
+
+template <>
+struct BinaryOpTypeName<BinaryOpType::kSub> {
+  static constexpr char kName[] = "sub.out";
+};
+
+template <>
+struct BinaryOpTypeName<BinaryOpType::kMul> {
+  static constexpr char kName[] = "mul.out";
+};
+
+template <>
+struct BinaryOpTypeName<BinaryOpType::kDiv> {
+  static constexpr char kName[] = "div.out";
+};
+
```
you don't need to do this. see existing example: https://github.com/pytorch/executorch/blob/c82a7df29e25b1fb348d8e833d82fe00d95ce45d/kernels/portable/cpu/op_rsub.cpp#L50-L55

the secret sauce is that the string literal has to be a static constexpr const char [] and then you can pass it to a `const char*` template argument directly.

kimishpatel replied:
Thanks. I was hoping you would point me to something better for this

kimishpatel commented on kernels/optimized/cpu/binary_ops.h:
```diff
@@ -191,14 +222,15 @@ std::array<int32_t, 3> inline get_normalized_tensor_size(
     Tensor& out,
-    const ElementwiseOptimizedPath selected_optimized_path) {
+    const ElementwiseOptimizedPath selected_optimized_path,
+    executorch::aten::optional<Scalar>& alpha = {}) {
```
yeah just realized that. not sure why it did not throw error in local build but different compile options i guess.

I just followed what i see elsewhere. Happy to switch to std::optional too which is what I guess is backing that but maybe for aten build it aliases c10:::optional? Let me check that first

swolchok replied:
c10::optional is gone

kimishpatel commented on kernels/optimized/cpu/binary_ops.h:
```diff
@@ -231,20 +276,21 @@ Tensor& handle_last_dim_broadcast_elementwise(
     Tensor& out,
-    const ElementwiseOptimizedPath selected_optimized_path) {
+    const ElementwiseOptimizedPath selected_optimized_path,
+    executorch::aten::optional<Scalar> alpha = {}) {
```
oh good call out. my bad

kimishpatel committed:
Update base for Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

swolchok commented on kernels/optimized/cpu/op_add.cpp:
```diff
@@ -140,41 +140,32 @@ Tensor& opt_add_out(
+      // This should really be resolved in handle_broadcast_elementwise.
+      // However, the current blocker is that handle_broadcast_elementwise tries
+      // to be agnostic of op. This should be fixed, likely by moving lambda
+      // creation to handle_broadcast_elementwise and it be aware of which op is
+      // being executed.
```
not sure I agree, but we can settle that on a review of a proposed change

swolchok commented on kernels/optimized/cpu/binary_ops.h:
```diff
@@ -219,9 +221,22 @@ Tensor& handle_last_dim_broadcast_elementwise(
-    executorch::vec::broadcasting_map_broadcast_last_dim<CTYPE, Op>(
-        vec_fun,
+  ET_SWITCH_REALB_TYPES(out_type, ctx, op_name, CTYPE, [&]() {
+    using Vec = executorch::vec::Vectorized<CTYPE>;
+    Vec alpha_val_vec;
+    if (alpha.has_value()) {
+      CTYPE alpha_val;
+      ET_KERNEL_CHECK(
+          ctx,
+          native::utils::extract_scalar(alpha.value(), &alpha_val),
+          InvalidArgument, );
+      alpha_val_vec = Vec(alpha_val);
+    }
+    auto vec_fun_alpha = [vec_fun, alpha_val_vec](const Vec& a, const Vec& b) {
+      return vec_fun(a, b, alpha_val_vec);
+    };
```
oh, I see, you're having problems with the lambda because of this part. you can solve this by factoring the code differently.

the end result at the callsite could look something like

```
auto broadcast_op_plan_opt = plan_broadcast_elementwise(...); // broadcast_op_plan is a struct containing all the stuff you work out that isn't dependent on the dtype, like lhs, rhs. it does ET_KERNEL_CHECKs intenrally and returns nullopt if they fail.
if (!broadcast_op_plan_opt) {
  // a check already failed
  return;
}
ET_SWITCH_REALB_TYPES(out_type, ctx, op_name, CTYPE, [&]() {
  auto alpha_val_vec_opt = extract_scalar_to_vector<CTYPE>(); // wrap up the bit that 
  if (!alpha_val_vec_opt) {
    // awkward that this only returns from the lambda, but this is a generic ET_KERNEL_CHECK problem
    return;
  }
  auto add_lambda = [alpha_val_vec = *alpha_val_vec_opt](auto x, auto y) {
        return y + alpha_val * x;
      };
  execute_broadcast_elementwise_plan<CTYPE>(*broadcast_op_plan_opt, add_lambda, ...);
});
```

disclaimer: this is off the top of my head and it may be possible to unify some of this stuff with dtype_util.h for further simplification, though dtype_util is mostly intended to cut size/build time of portable ops

kimishpatel replied:
Good point. Let me see if I dont run into other issues to enable such refactor.

kimishpatel replied:
Ok so I looked refactor required. I think it is doable at the cost of moving `ET_SWITCH_REALB_TYPES` macros to the callsite in respective ops. Downside here is that now if you enable new dtype for optimized path, you have to change all the callsites.

So I am not fully convinced that it is better go down that route. But want to see whats your reasoning.

swolchok replied:
> you have to change all the callsites.

that's just a matter of typing, right? if you plan to do it (I suppose optimizing Half/BFloat16 should be on our TODO list if the hardware supports the relevant instructions) and you really don't want to change 4-5 files later (you'll have to change them anyway for specifically Half/BFloat16 because there are opt-outs), you could always `#define ET_SWITCH_OPTIMIZED_ELEMENTWISE_BROADCAST_OP_TYPES ET_SWITCH_REALB_TYPEs` pre-emptively.

kimishpatel replied:
Ok thats fair. But is your reasoning for this change simpler code or you see perf impact.

I am not too stuck to it, so I will just go ahead and do it but wanted to understand your reasoning

swolchok replied:
simpler less repetitive code

kimishpatel replied:
ok will make the change but this will likely marginally increase size since now the whole `handle_broadcast_elementwise` function is dtype specialized

swolchok reviewed (changes requested):
just re-request review if my suggestion is bad :)

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

swolchok commented on kernels/test/op_add_test.cpp:
```diff
@@ -112,6 +112,125 @@ class OpAddOutKernelTest : public OperatorTest {
+  void test_broadcast_4D() {
+    TensorFactory<DTYPE> tf_a;
+
+    Tensor a = tf_a.make(
+        {2, 2, 3, 5},
+        /*data=*/{1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15,
+                  16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30,
+                  31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45,
+                  46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60});
+    Tensor b = tf_a.make(
+        {2, 1, 3, 5},
+        /*data=*/{1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15,
+                  16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30});
```
nit: it would probably be more reviewable to fill these programmatically, such as with [std::iota](https://en.cppreference.com/w/cpp/algorithm/iota), but certainly not blocking

swolchok commented on kernels/test/op_add_test.cpp:
```diff
@@ -112,6 +112,125 @@ class OpAddOutKernelTest : public OperatorTest {
+    Tensor out = tf_a.zeros({2, 2, 3, 5});
+    Tensor expected = tf_a.make(
+        {2, 2, 3, 5},
+        /*data=*/{2,  4,  6,  8,  10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30,
+                  17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45,
+                  47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75,
+                  62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90});
```
ditto programmatic fill

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491814).

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491814).

kimishpatel committed:
Update base for Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491814).

kimishpatel committed:
Update base for Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel committed:
Merge branch 'main' into gh/kimishpatel/154/head

kimishpatel committed:
Update base for Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491814](https://our.internmc.facebook.com/intern/diff/D69491814)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491814).
